<a href="https://colab.research.google.com/github/TiagoHarari/DataAnalysis_course/blob/main/1-indexing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/institutohumai/cursos-python/blob/master/AnalisisDeDatos/1_Indexing/Indexing.ipynb"> <img src='https://colab.research.google.com/assets/colab-badge.svg' /> </a>
<div align="center"> Recordá abrir en una nueva pestaña </div>

In [ ]:
# Siempre al principio, importamos las librerías.
import pandas as pd
import numpy as np

# Indexación y Agregación

Tabla de Contenidos

    I. Análisis de datos con Pandas
    II. Los objetos fundamentales de Pandas
        I. Series
        II. DataFrames
        III. Índices
    III. Exploración
        I. Filtrando un DataFrame (Indexing)
            I. Boolean Indexing
                I. Máscara booleana
                II. Máscara booleana con muchas condiciones
            II. Boolean indexing con query()
            III. Fancy Indexing
        II. Funciones de Agregación
    IV. Otros análisis descriptivos
        I. Para las variables numéricas
        II. Para las variables categóricas
        III. Ordenar por columnas y limitar la cantidad de resultados
    V. Anexo: volviendo al tema de la vectorización

## Exploración

Vamos a analizar datos de una fuente real. Los ingresos de los funcionarios son información pública que se libera anualmente en el <a href='https://data.buenosaires.gob.ar/dataset/sueldo-funcionarios'>portal de datos abiertos</a> de GCBA.  

En general los 4 primeros pasos para analizar un data set son:
1. Leerlo
2. Consultar cuáles son las columnas
3. Extraer una muestra
4. Verificar cuántos registros tiene

## 1- Para leer el data set usamos la función de pandas read_csv

Con esta función podemos leer archivos que estén en una url pública o en una ubicación del disco accesible desde la Jupyter Notebook.

In [ ]:
df = pd.read_csv('http://cdn.buenosaires.gob.ar/datosabiertos/datasets/sueldo-funcionarios/sueldo_funcionarios_2019.csv')

## 2- Consultamos las columnas


In [ ]:
df.columns

Index(['cuil', 'anio', 'mes', 'funcionario_apellido', 'funcionario_nombre',
       'reparticion', 'asignacion_por_cargo_i', 'aguinaldo_ii',
       'total_salario_bruto_i_+_ii', 'observaciones'],
      dtype='object')

In [ ]:
df.head()

cuil  anio  mes funcionario_apellido funcionario_nombre  \
0  20-17692128-6  2019    1    RODRIGUEZ LARRETA    HORACIO ANTONIO   
1  20-17735449-0  2019    1             SANTILLI        DIEGO CESAR   
2  27-24483014-0  2019    1                ACUÑA      MARIA SOLEDAD   
3  20-13872301-2  2019    1             ASTARLOA      GABRIEL MARIA   
4  20-25641207-2  2019    1             AVOGADRO       ENRIQUE LUIS   

                                        reparticion  asignacion_por_cargo_i  \
0                                  Jefe de Gobierno               197745.80   
1                          Vicejefatura de Gobierno               197745.80   
2              Ministerio de Educación e Innovación               224516.62   
3  Procuración General de la Ciudad de Buenos Aires               224516.62   
4                             Ministerio de Cultura               224516.62   

   aguinaldo_ii  total_salario_bruto_i_+_ii observaciones  
0           0.0                   197745.80           NaN  
1           0.0                   197745.80           NaN  
2           0.0                   224516.62           NaN  
3           0.0                   224516.62           NaN  
4           0.0                   224516.62           NaN

## 3- Extraemos una muestra

In [ ]:
df.sample(5)

cuil  anio  mes funcionario_apellido funcionario_nombre  \
96   20-17735449-0  2019    4             SANTILLI        DIEGO CESAR   
259  20-14699669-9  2019    9               MOCCIA             FRANCO   
277  27-26690980-8  2019    9               UHALDE        MARIA PAULA   
38   27-13092400-5  2019    2                FREDA     MONICA BEATRIZ   
213  20-26735689-1  2019    7               MENDEZ          JUAN JOSE   

                                          reparticion  asignacion_por_cargo_i  \
96                           Vicejefatura de Gobierno               217520.38   
259      Ministerio de Desarrollo Urbano y Transporte               275089.75   
277          SECR Cultura Ciudadana y Función Pública               249972.87   
38   Sindicatura General de la Ciudad de Buenos Aires               224516.62   
213                                   SECR Transporte               239470.36   

     aguinaldo_ii  total_salario_bruto_i_+_ii observaciones  
96            0.0                   217520.38           NaN  
259           0.0                   275089.75           NaN  
277           0.0                   249972.87           NaN  
38            0.0                   224516.62           NaN  
213           0.0                   239470.36           NaN

## 4- Consultamos la cantidad de filas y de columnas

In [ ]:
# La propiedad shape nos devuelve una tupla (filas,columnas)
df.shape

(385, 10)

## Vectorización con Pandas

Pandas es una de las librerías de Python más usadas para análisis de datos. El nombre pandas viene de "Panel Data Analysis" y su funcionalidad permite hacer operaciones sobre datos que se encuentran en memoria de manera eficiente.

Pandas es útil para trabajar sobre datos tabulares, con dos condiciones importantes:

I. Los datos se encuentran enteramente en la memoria RAM. Con lo cual, el tamaño de los datos que podemos manipular está limitado por el hardware. Como regla de pulgar, es una buena práctica no ocupar más de 1/3 de la memoria RAM de nuestro dispositivo con el dataset. Así, si estamos trabajando localmente en una notebook con 8GB de memoria RAM no es recomendable procesar datasets de más de 2.33GB.

II. En pandas, las operaciones sobre filas y columnas son, en general, eficientes porque se hacen de forma "vectorizada". En realidad esta optimización, se hace desde numpy, una librería para realizar operaciones matemáticas que se utilizó a su vez para escribir pandas.

Las operaciones vectorizadas son las que se realizan en bloque en vez de caso por caso. Las computadoras de hoy tienen la capacidad de recibir muchas instrucciones juntas y procesar varias de ellas a la vez. Por ejemplo, si nuestro hardware tiene la capacidad de procesar 4 operaciones juntas, el resultado de vectorizar una operación matemática es el siguiente:

<img src = 'https://storage.googleapis.com/humai-datasets/aws_s3/analisis_de_datos/1_Indexing/vectorizacion.png' />


En el primer caso hay que hacer 5 operaciones y en el segundo caso sólo dos.

Es importante entender, entonces, que Pandas trabaja de esta manera y que por eso es una de las herramientas más elegidas para manipular datos en memoria.


## Los objetos fundamentales de Pandas

### Series

Las series son "columnas" que de una tabla que están asociadas a un índice y a un nombre. Igual que una lista común de Python es una secuencia de elementos ordenados, pero a diferencia de la lista está asociada a más información.

In [ ]:
# Las series se pueden crear a partir de una lista
serie = pd.Series(['a','b','c'])

In [ ]:
# Propiedades importantes de las series
print('Tipo de objetos que tiene ', serie.dtype)
print('Nombre ', serie.name)
print('Index ',serie.index)
print('Valores ',serie.values)


Tipo de objetos que tiene  object
Nombre  None
Index  RangeIndex(start=0, stop=3, step=1)
Valores  ['a' 'b' 'c']


## DataFrames

Los DataFrames son "tablas", compuestas por varias "columnas" o series que comparten todas un mismo índice. En general los DataFrames se crean a partir de leer tablas de archivos (pueden ser en formato json o csv) pero a veces también se crean a partir de listas de diccionarios o de otras maneras.

Los DataFrames tienen un objeto Index que describe los nombres de columnas y otro objeto Index que describen los nombres de las filas.

In [ ]:
# Leemos un dataset público
df = pd.read_csv('http://cdn.buenosaires.gob.ar/datosabiertos/datasets/sueldo-funcionarios/sueldo_funcionarios_2019.csv')

In [ ]:
# Propiedades importantes de los dataframes
print('Columnas ', df.columns)
print('Index ', df.index)
print('Dimensiones ',df.shape)


Columnas  Index(['cuil', 'anio', 'mes', 'funcionario_apellido', 'funcionario_nombre',
       'reparticion', 'asignacion_por_cargo_i', 'aguinaldo_ii',
       'total_salario_bruto_i_+_ii', 'observaciones'],
      dtype='object')
Index  RangeIndex(start=0, stop=385, step=1)
Dimensiones  (385, 10)


In [ ]:
# Consultar las primeras filas
df.head()

cuil  anio  mes funcionario_apellido funcionario_nombre  \
0  20-17692128-6  2019    1    RODRIGUEZ LARRETA    HORACIO ANTONIO   
1  20-17735449-0  2019    1             SANTILLI        DIEGO CESAR   
2  27-24483014-0  2019    1                ACUÑA      MARIA SOLEDAD   
3  20-13872301-2  2019    1             ASTARLOA      GABRIEL MARIA   
4  20-25641207-2  2019    1             AVOGADRO       ENRIQUE LUIS   

                                        reparticion  asignacion_por_cargo_i  \
0                                  Jefe de Gobierno               197745.80   
1                          Vicejefatura de Gobierno               197745.80   
2              Ministerio de Educación e Innovación               224516.62   
3  Procuración General de la Ciudad de Buenos Aires               224516.62   
4                             Ministerio de Cultura               224516.62   

   aguinaldo_ii  total_salario_bruto_i_+_ii observaciones  
0           0.0                   197745.80           NaN  
1           0.0                   197745.80           NaN  
2           0.0                   224516.62           NaN  
3           0.0                   224516.62           NaN  
4           0.0                   224516.62           NaN

Si queremos extraer una serie del DataFrame, podemos hacerlo de la misma forma en que extraemos un valor de un diccionario.



In [ ]:
serie_mes = df['mes']

In [ ]:
type(serie_mes)

pandas.core.series.Series

## Índices

Los índices acompañan a las series y a los Data Frames. Son conjuntos ordenados e inmutables de elementos

In [ ]:
df.index

RangeIndex(start=0, stop=385, step=1)

In [ ]:
df.columns

Index(['cuil', 'anio', 'mes', 'funcionario_apellido', 'funcionario_nombre',
       'reparticion', 'asignacion_por_cargo_i', 'aguinaldo_ii',
       'total_salario_bruto_i_+_ii', 'observaciones'],
      dtype='object')

In [ ]:
ind = pd.Index([2, 3, 5, 7, 11])
ind

Index([2, 3, 5, 7, 11], dtype='int64')

### Ejercicio
Exploren el dataset público que se encuentra en la siguiente url: https://storage.googleapis.com/humai-datasets/aws_s3/analisis_de_datos/1_Indexing/titanic.csv ¿De qué se trata? ¿Cuántas filas tiene? ¿Cuántas columnas? Al leerlo, pueden almacenarlo en la variable df_titanic.

In [ ]:
df_titanic = pd.read_csv("https://storage.googleapis.com/humai-datasets/aws_s3/analisis_de_datos/1_Indexing/titanic.csv", sep = "|")

In [ ]:
df_titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
df_titanic.shape #891 filas y 12 variables (columnas, series)

(891, 12)

## Filtrando un DataFrame (Indexing)

Hay muchas técnicas para filtrar un DataFrame. Podemos querer filtrar por columnas o por filas, por posición o por nombre. También podemos querer filtrar por condiciones que se cumplen o no. Cuando no queremos filtrar sobre una dimensión (filas o columnas) usamos ":" para seleccionar todo.


<img src='https://github.com/institutohumai/cursos-python/blob/master/AnalisisDeDatos/1_Indexing/img/indexing.png?raw=1' style='height:350px' />



### Boolean Indexing

Supongamos que queremos tomar el dataset de funcionarios y quedarnos únicamente con los que pertenecen al Ministerio de Cultura.
Para eso lo que hacemos es indexar al DataFrame por una condición booleana. Eso implica que debemos crear una serie compuesta por valores True y False para aplicarla como índice a las filas.

Los operadores que sirven para evaluar condiciones sobre las series son:


| S  | Descripción   | S  | Descripción   |   |
|----|---------------|----|---------------|---|
| >= | Mayor o Igual | <= | Menor o Igual |   |
| == | Igual         | != | Distinto      |   |
| >  | Mayor         | <  | Menor         |   |

#### Máscara booleana

Veamos lo que pasa cuando le aplicamos a una serie una condición que devuelve un booleano

In [ ]:
df['anio'] != 2019

0      False
1      False
2      False
3      False
4      False
       ...  
380    False
381    False
382    False
383    False
384    False
Name: anio, Length: 385, dtype: bool

In [ ]:
mascara_booleana = df['anio'] != 2019

Nos devuelve una serie de la misma longitud que la original y que contiene sólo valores True o False.

In [ ]:
type(mascara_booleana)

pandas.core.series.Series

In [ ]:
mascara_booleana.shape

(385,)

In [ ]:
mascara_booleana.dtype

dtype('bool')

Ahora seleccionemos entonces, los registros que corresponden al Ministerio de Cultura.

In [ ]:
df_min_cul = df.loc[df['reparticion'] == 'Ministerio de Cultura',:] #USAMOS LOC[ condicional para las filas del min. cul. , : (todas las columnas)]

In [ ]:
# Veamos la cantidad de casos
df_min_cul.shape

(12, 10)

Algo que puede llegar a confundir sobre el Indexing en Pandas es que en algunos casos se puede ser menos explícito a la hora de filtrar. Por ejemplo si ponemos una condición Booleana, pandas asume que el tipo de indexing es loc y que el filtro es sobre las filas y no sobre las columnas:


In [ ]:
df_min_cul = df[df['reparticion'] == 'Ministerio de Cultura']

In [ ]:
df_min_cul.shape

(12, 10)

Probemos con otra condición.

### Ejercicio

Traer todos los sueldos de la segunda mitad del año...

In [ ]:
df_segunda_mitad = df[df["mes"] > 6]
print(df_segunda_mitad)

              cuil  anio  mes funcionario_apellido funcionario_nombre  \
188  20-17692128-6  2019    7    RODRIGUEZ LARRETA    HORACIO ANTONIO   
189  20-17735449-0  2019    7             SANTILLI        DIEGO CESAR   
190  27-24483014-0  2019    7                ACUÑA      MARIA SOLEDAD   
191  20-13872301-2  2019    7             ASTARLOA      GABRIEL MARIA   
192  20-25641207-2  2019    7             AVOGADRO       ENRIQUE LUIS   
..             ...   ...  ...                  ...                ...   
380  20-26735689-1  2019   12               MENDEZ          JUAN JOSE   
381  20-23864572-8  2019   12             STRAFACE     FERNANDO DIEGO   
382  20-21981279-6  2019   12        SCRENCI SILVA        BRUNO GUIDO   
383  20-30593774-7  2019   12           AVELLANEDA   PATRICIO IGNACIO   
384  27-27011181-0  2019   12            FERNANDEZ     KARINA BEATRIZ   

                                           reparticion  \
188                                   Jefe de Gobierno   
189    

### Ejercicio
Volviendo al DataFrame del Titanic ¿Cuántos pasajeros sobrevivieron y cuántos no? ¿Cuántos pagaron una tarifa menor a 25?

In [ ]:
df_titanic.head() #visualizamospara ver variables

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
#0 es que no sobrevieron, 1 es que si sobrevivieron
sobrevivientes = df_titanic[df_titanic["Survived"] == 1]
muertes = df_titanic[df_titanic["Survived"] == 0]

In [ ]:
print(sobrevivientes.shape)
print(muertes.shape)

(342, 12)
(549, 12)


In [ ]:
print(df_titanic["Survived"].value_counts()) #print(de la variable survived del df de titanic --> contar valores)

Survived
0    549
1    342
Name: count, dtype: int64


In [ ]:
df_tarifa_baja = df_titanic[df_titanic["Fare"] < 25]
df_tarifa_baja.shape

(557, 12)

#### Máscara booleana con muchas condiciones

Ahora tratemos de filtrar el dataset por dos condiciones: por ejemplo tomar los sueldos de abril de la secretaria de innovación.
Para eso tenemos que combinar dos máscaras booleanas con una condición.

| S | Descripcion        | S  | Descripcion |   |
|---|--------------------|----|-------------|---|
| & | AND (y)            | \| | OR (o)      |   |
| ^ | XOR (o exclusivo)  | ~  | NOT (no)    |   |



Por ejemplo: seleccionemos los casos donde o bien se haya cobrado aguinaldo o bien el salario total haya sido mayor que 240.000, pero no las dos cosas.


In [ ]:
df[(df['total_salario_bruto_i_+_ii'] > 240000) ^ (df['aguinaldo_ii'] > 0)] #usamos un XOR porque es o una o la otra

cuil  anio  mes funcionario_apellido funcionario_nombre  \
55   20-26781618-3  2019    2                LARRE       PEDRO ANDRES   
97   27-24483014-0  2019    4                ACUÑA      MARIA SOLEDAD   
98   20-13872301-2  2019    4             ASTARLOA      GABRIEL MARIA   
99   20-25641207-2  2019    4             AVOGADRO       ENRIQUE LUIS   
100  27-13221055-7  2019    4            BOU PEREZ          ANA MARIA   
..             ...   ...  ...                  ...                ...   
360  20-28908968-4  2019   12       COELHO CHICANO          CHRISTIAN   
361  20-28908968-4  2019   12       COELHO CHICANO          CHRISTIAN   
362  20-24424714-9  2019   12         D'ALESSANDRO     MARCELO SILVIO   
363  20-24424714-9  2019   12         D'ALESSANDRO     MARCELO SILVIO   
384  27-27011181-0  2019   12            FERNANDEZ     KARINA BEATRIZ   

                                          reparticion  asignacion_por_cargo_i  \
55              SECR Ciencia, Tecnologia e Innovacion               204017.27   
97               Ministerio de Educación e Innovación               249661.60   
98   Procuración General de la Ciudad de Buenos Aires               249661.60   
99                              Ministerio de Cultura               249661.60   
100                               Ministerio de Salud               249661.60   
..                                                ...                     ...   
360                                     SS Contenidos                74991.86   
361                                    SECR de Medios               187539.45   
362                         SECR Justicia y Seguridad                74991.86   
363                         SECR Justicia y Seguridad               187539.45   
384                         SECR Asuntos Estratégicos               187539.45   

     aguinaldo_ii  total_salario_bruto_i_+_ii        observaciones  
55       34002.88                   238020.15       baja 28/2/2019  
97           0.00                   249661.60                  NaN  
98           0.00                   249661.60                  NaN  
99           0.00                   249661.60                  NaN  
100          0.00                   249661.60                  NaN  
..            ...                         ...                  ...  
360     110404.68                   185396.54         baja al 9/12  
361      15628.29                   203167.74  alta desde el 10/12  
362     110404.68                   185396.54         baja al 9/12  
363      15628.29                   203167.74  alta desde el 10/12  
384      15628.29                   203167.74  alta desde el 10/12  

[170 rows x 10 columns]

Ahora veamos los sueldos de febrero de la SECR Ciencia, Tecnologia e Innovacion.

In [ ]:
df[(df['mes'] == 2) & (df['reparticion'] == 'SECR Ciencia, Tecnologia e Innovacion')]

cuil  anio  mes funcionario_apellido funcionario_nombre  \
55  20-26781618-3  2019    2                LARRE       PEDRO ANDRES   

                              reparticion  asignacion_por_cargo_i  \
55  SECR Ciencia, Tecnologia e Innovacion               204017.27   

    aguinaldo_ii  total_salario_bruto_i_+_ii   observaciones  
55      34002.88                   238020.15  baja 28/2/2019

### Boolean indexing con query()

La sintaxis que se utiliza para hacer Boolean indexing es un poco repetitiva. Noten que filtrar (aún en su expresión más corta sin loc ni especificar filas o columnas) implica ESCRIBIR DOS VECES el nombre del dataset. Para crear un shortcut, Pandas ofrece la función .query()



In [ ]:
#usamos query y condiciones
df_cult = df.query('reparticion == "Ministerio de Cultura"')
df_cult.head()

cuil  anio  mes funcionario_apellido funcionario_nombre  \
4    20-25641207-2  2019    1             AVOGADRO       ENRIQUE LUIS   
36   20-25641207-2  2019    2             AVOGADRO       ENRIQUE LUIS   
68   20-25641207-2  2019    3             AVOGADRO       ENRIQUE LUIS   
99   20-25641207-2  2019    4             AVOGADRO       ENRIQUE LUIS   
130  20-25641207-2  2019    5             AVOGADRO       ENRIQUE LUIS   

               reparticion  asignacion_por_cargo_i  aguinaldo_ii  \
4    Ministerio de Cultura               224516.62           0.0   
36   Ministerio de Cultura               224516.62           0.0   
68   Ministerio de Cultura               231167.76           0.0   
99   Ministerio de Cultura               249661.60           0.0   
130  Ministerio de Cultura               249661.60           0.0   

     total_salario_bruto_i_+_ii observaciones  
4                     224516.62           NaN  
36                    224516.62           NaN  
68                    231167.76           NaN  
99                    249661.60           NaN  
130                   249661.60           NaN

También se puede hacer query sobre múltiples condiciones.

In [ ]:
#que hayan cobrado y que hayan cobrado aguinaldo
df2 = df.query('asignacion_por_cargo_i > 240000 & aguinaldo_ii > 0')
df2.head()

cuil  anio  mes funcionario_apellido funcionario_nombre  \
159  27-24483014-0  2019    6                ACUÑA      MARIA SOLEDAD   
160  20-13872301-2  2019    6             ASTARLOA      GABRIEL MARIA   
161  20-25641207-2  2019    6             AVOGADRO       ENRIQUE LUIS   
162  27-13221055-7  2019    6            BOU PEREZ          ANA MARIA   
163  27-13092400-5  2019    6                FREDA     MONICA BEATRIZ   

                                          reparticion  asignacion_por_cargo_i  \
159              Ministerio de Educación e Innovación                249661.6   
160  Procuración General de la Ciudad de Buenos Aires                249661.6   
161                             Ministerio de Cultura                249661.6   
162                               Ministerio de Salud                249661.6   
163  Sindicatura General de la Ciudad de Buenos Aires                249661.6   

     aguinaldo_ii  total_salario_bruto_i_+_ii observaciones  
159     124830.80                   374492.40           NaN  
160     165196.76                   414858.36           NaN  
161     124830.80                   374492.40           NaN  
162     124830.80                   374492.40           NaN  
163     124830.80                   374492.40           NaN

In [ ]:
df2.shape

(34, 10)

### Ejercicio: Piensen cómo traducir a la sintaxis de query, estas consultas que ya hicimos:

In [ ]:
# df_sem2 = df[df['mes'] > 6]
df_sem2 = df.query('mes > 6')
df_sem2.head()


cuil  anio  mes funcionario_apellido funcionario_nombre  \
188  20-17692128-6  2019    7    RODRIGUEZ LARRETA    HORACIO ANTONIO   
189  20-17735449-0  2019    7             SANTILLI        DIEGO CESAR   
190  27-24483014-0  2019    7                ACUÑA      MARIA SOLEDAD   
191  20-13872301-2  2019    7             ASTARLOA      GABRIEL MARIA   
192  20-25641207-2  2019    7             AVOGADRO       ENRIQUE LUIS   

                                          reparticion  asignacion_por_cargo_i  \
188                                  Jefe de Gobierno               228396.40   
189                          Vicejefatura de Gobierno               228396.40   
190              Ministerio de Educación e Innovación               263531.98   
191  Procuración General de la Ciudad de Buenos Aires               263531.98   
192                             Ministerio de Cultura               263531.98   

     aguinaldo_ii  total_salario_bruto_i_+_ii observaciones  
188           0.0                   228396.40           NaN  
189           0.0                   228396.40           NaN  
190           0.0                   263531.98           NaN  
191           0.0                   263531.98           NaN  
192           0.0                   263531.98           NaN

In [ ]:
# df[(df['mes'] == 2) & (df['reparticion'] == 'SECR Ciencia, Tecnologia e Innovacion')]
df_feb_SECR = df.query('mes == 2 & reparticion == "SECR Ciencia, Tecnologia e Innovacion"')
df_feb_SECR.head()

cuil  anio  mes funcionario_apellido funcionario_nombre  \
55  20-26781618-3  2019    2                LARRE       PEDRO ANDRES   

                              reparticion  asignacion_por_cargo_i  \
55  SECR Ciencia, Tecnologia e Innovacion               204017.27   

    aguinaldo_ii  total_salario_bruto_i_+_ii   observaciones  
55      34002.88                   238020.15  baja 28/2/2019

### Fancy Indexing

Ahora vamos a quedarnos con un subconjunto de columnas del DataFrame.

In [ ]:
df_view = df.loc[:,['anio','mes']]
#selecciono todas las filas , y las columna anio y mes.

In [ ]:
df_view.shape

(385, 2)

Existe una forma menos explícita de hacer esta misma operación. Si pasamos una lista al indexing, pandas asume que el tipo de indexing es loc y que el filtro es sobre las columnas y no las filas:

In [ ]:
df_view = df[['anio','mes']]
# es lo mismo que antes solo que sin utilizar loc

In [ ]:
df_view.shape

(385, 2)

Fíjense lo que pasa si tratamos de acceder a filas utilizando una lista de nombres, en este caso [0,1].

Nos da un error porque cuando pasamos únicamente una lista al indexing, pandas asume que queremos un set de columnas y si los nombres no existen, da error. La forma correcta de hacerlo es pasar una lista de índices y explicitar que vamos a indizar con loc y que seleccionamos todas las columnas.

In [ ]:
# Correcto
df_view = df.loc[[3,8],:]
#filas 3 y 8. Todas las columnas

In [ ]:
df_view

cuil  anio  mes funcionario_apellido funcionario_nombre  \
3  20-13872301-2  2019    1             ASTARLOA      GABRIEL MARIA   
8  20-22293873-3  2019    1               MIGUEL       FELIPE OSCAR   

                                        reparticion  asignacion_por_cargo_i  \
3  Procuración General de la Ciudad de Buenos Aires               224516.62   
8                 Jefatura de Gabinete de Ministros               224516.62   

   aguinaldo_ii  total_salario_bruto_i_+_ii observaciones  
3           0.0                   224516.62           NaN  
8           0.0                   224516.62           NaN

### Ejercicio. Volviendo al ejemplo del titanic...

1) ¿Cuántos hombres y mujeres sobrevivieron?

2) ¿Cuántos menores de 18 años había? ¿Cuántos sobrevivieron?

3) Seleccionen únicamente las columnas Sex y Survived y almacenenlas en un nuevo DataFrame que se llame df_titanic_subset.


In [ ]:
df_titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
# 1) ¿Cuántos mujeres sobrevivieron?
df_titanic.query('Sex == "female" & Survived == 1')


PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
874          875         1       2   
875          876         1       3   
879          880         1       1   
880          881         1       2   
887          888         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
874              Abelson, Mrs. Samuel (Hannah Wizosky)  female  28.0      1   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   

     Parch            Ticket     Fare Cabin Embarked  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
..     ...               ...      ...   ...      ...  
874      0         P/PP 3381  24.0000   NaN        C  
875      0              2667   7.2250   NaN        C  
879      1             11767  83.1583   C50        C  
880      1            230433  26.0000   NaN        S  
887      0            112053  30.0000   B42        S  

[233 rows x 12 columns]

In [ ]:
#1) ¿Cuántos hombres sobrevivieron?
df_titanic.query('Sex == "male" & Survived ==  1')

PassengerId  Survived  Pclass                             Name   Sex  \
17            18         1       2     Williams, Mr. Charles Eugene  male   
21            22         1       2            Beesley, Mr. Lawrence  male   
23            24         1       1     Sloper, Mr. William Thompson  male   
36            37         1       3                 Mamee, Mr. Hanna  male   
55            56         1       1                Woolner, Mr. Hugh  male   
..           ...       ...     ...                              ...   ...   
838          839         1       3                  Chip, Mr. Chang  male   
839          840         1       1             Marechal, Mr. Pierre  male   
857          858         1       1           Daly, Mr. Peter Denis   male   
869          870         1       3  Johnson, Master. Harold Theodor  male   
889          890         1       1            Behr, Mr. Karl Howell  male   

      Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
17    NaN      0      0  244373  13.0000   NaN        S  
21   34.0      0      0  248698  13.0000   D56        S  
23   28.0      0      0  113788  35.5000    A6        S  
36    NaN      0      0    2677   7.2292   NaN        C  
55    NaN      0      0   19947  35.5000   C52        S  
..    ...    ...    ...     ...      ...   ...      ...  
838  32.0      0      0    1601  56.4958   NaN        S  
839   NaN      0      0   11774  29.7000   C47        C  
857  51.0      0      0  113055  26.5500   E17        S  
869   4.0      1      1  347742  11.1333   NaN        S  
889  26.0      0      0  111369  30.0000  C148        C  

[109 rows x 12 columns]

In [ ]:
#¿Cuántos menores de 18 años había? ¿Cuántos sobrevivieron?
df_titanic.query('Age < 18 & Survived == 1')

PassengerId  Survived  Pclass                                      Name  \
9             10         1       2       Nasser, Mrs. Nicholas (Adele Achem)   
10            11         1       3           Sandstrom, Miss. Marguerite Rut   
22            23         1       3               McGowan, Miss. Anna "Annie"   
39            40         1       3               Nicola-Yarred, Miss. Jamila   
43            44         1       2  Laroche, Miss. Simonne Marie Anne Andree   
..           ...       ...     ...                                       ...   
830          831         1       3   Yasbeck, Mrs. Antoni (Selini Alexander)   
831          832         1       2           Richards, Master. George Sibley   
853          854         1       1                 Lines, Miss. Mary Conover   
869          870         1       3           Johnson, Master. Harold Theodor   
875          876         1       3          Najib, Miss. Adele Kiamie "Jane"   

        Sex    Age  SibSp  Parch         Ticket     Fare Cabin Embarked  
9    female  14.00      1      0         237736  30.0708   NaN        C  
10   female   4.00      1      1        PP 9549  16.7000    G6        S  
22   female  15.00      0      0         330923   8.0292   NaN        Q  
39   female  14.00      1      0           2651  11.2417   NaN        C  
43   female   3.00      1      2  SC/Paris 2123  41.5792   NaN        C  
..      ...    ...    ...    ...            ...      ...   ...      ...  
830  female  15.00      1      0           2659  14.4542   NaN        C  
831    male   0.83      1      1          29106  18.7500   NaN        S  
853  female  16.00      0      1       PC 17592  39.4000   D28        S  
869    male   4.00      1      1         347742  11.1333   NaN        S  
875  female  15.00      0      0           2667   7.2250   NaN        C  

[61 rows x 12 columns]

In [ ]:
#3)Seleccionen únicamente las columnas Sex y Survived y almacenenlas en un nuevo DataFrame que se llame df_titanic_subset.
df_titanic_subset = df_titanic[["Sex", "Survived"]] #este metodo es mas facil

In [ ]:
df_titanic_subset1 = df_titanic.loc[:, ["Sex", "Survived"]]

In [ ]:
df_titanic_subset1.shape

(891, 2)

In [ ]:
#df_titanic_subset.head()
df_titanic_subset.head()


Sex  Survived
0    male         0
1  female         1
2  female         1
3  female         1
4    male         0

In [ ]:
df_titanic_subset.shape

(891, 2)

## Funciones de Agregación

Utilizando Pandas podemos aplicar funciones a nivel de columna. Algunas funciones predefinidas son la media, el desvío estándar y la sumatoria, el valor máximo y el mínimo.

Algunas de las funciones de agregación más comunes son:

<ul>
    <li>min</li>
    <li>max</li>
    <li>count</li>
    <li>sum</li>
    <li>prod</li>
    <li>mean</li>
    <li>median</li>
    <li>mode</li>
    <li>std</li>
    <li>var</li>
</ul>




In [ ]:
df['mes'].max()

12

In [ ]:
df['asignacion_por_cargo_i'].mean()

234234.36800000002

In [ ]:
df['asignacion_por_cargo_i'].std()

35043.160084661766

In [ ]:
df['total_salario_bruto_i_+_ii'].sum()

97988834.36000001

Podemos combinar los filtros que vimos antes con las funciones de agregación para responder preguntas cómo ¿Cuál fue en gasto en asignaciones de funcionarios para la Secretaría de Medios 2019? ¿Y para la de Justicia y Seguridad?

In [ ]:
#gasto de salarios para la secretaria de medios
df[df['reparticion'] == 'SECR de Medios']['total_salario_bruto_i_+_ii'].sum()

3232402.25

In [ ]:
#gasto de salarios para la secretaria de Justicia y Seguridad
df[df['reparticion'] == 'SECR Justicia y Seguridad']['total_salario_bruto_i_+_ii'].sum()

3029551.7300000004

Ahora respondamos algunas preguntas: ¿Quién o quiénes del dataset cobran el salario más alto? ¿Y el más bajo?

In [ ]:
df[df['total_salario_bruto_i_+_ii'] == df['total_salario_bruto_i_+_ii'].max()]

cuil  anio  mes funcionario_apellido funcionario_nombre  \
374  20-13872301-2  2019   12             ASTARLOA      GABRIEL MARIA   

                                          reparticion  asignacion_por_cargo_i  \
374  Procuración General de la Ciudad de Buenos Aires               275089.75   

     aguinaldo_ii  total_salario_bruto_i_+_ii observaciones  
374     170855.56                   445945.31           NaN

In [ ]:
df[df['total_salario_bruto_i_+_ii'] == df['total_salario_bruto_i_+_ii'].min()]

cuil  anio  mes funcionario_apellido funcionario_nombre  \
344  27-30744939-6  2019   12              FERRERO           GENOVEVA   
348  20-22709722-2  2019   12                LOPEZ             MATIAS   
354  20-20008464-1  2019   12               GIUSTI          JOSE LUIS   
358  20-31164337-2  2019   12         DI BENEDETTO           FEDERICO   
360  20-28908968-4  2019   12       COELHO CHICANO          CHRISTIAN   
362  20-24424714-9  2019   12         D'ALESSANDRO     MARCELO SILVIO   

                                           reparticion  \
344     SECR Administración de Seguridad y Emergencias   
348                          SECR Desarrollo Ciudadano   
354  UPE Transferencia de Funciones y Facultades en...   
358                                    SS Comunicacion   
360                                      SS Contenidos   
362                          SECR Justicia y Seguridad   

     asignacion_por_cargo_i  aguinaldo_ii  total_salario_bruto_i_+_ii  \
344                74991.86     110404.68                   185396.54   
348                74991.86     110404.68                   185396.54   
354                74991.86     110404.68                   185396.54   
358                74991.86     110404.68                   185396.54   
360                74991.86     110404.68                   185396.54   
362                74991.86     110404.68                   185396.54   

    observaciones  
344  baja al 9/12  
348  baja al 9/12  
354  baja al 9/12  
358  baja al 9/12  
360  baja al 9/12  
362  baja al 9/12

In [ ]:
ferrero = df.query('funcionario_apellido == "FERRERO"')
ferrero

cuil  anio  mes funcionario_apellido funcionario_nombre  \
21   27-30744939-6  2019    1              FERRERO           GENOVEVA   
53   27-30744939-6  2019    2              FERRERO           GENOVEVA   
85   27-30744939-6  2019    3              FERRERO           GENOVEVA   
116  27-30744939-6  2019    4              FERRERO           GENOVEVA   
147  27-30744939-6  2019    5              FERRERO           GENOVEVA   
178  27-30744939-6  2019    6              FERRERO           GENOVEVA   
209  27-30744939-6  2019    7              FERRERO           GENOVEVA   
240  27-30744939-6  2019    8              FERRERO           GENOVEVA   
271  27-30744939-6  2019    9              FERRERO           GENOVEVA   
302  27-30744939-6  2019   10              FERRERO           GENOVEVA   
333  27-30744939-6  2019   11              FERRERO           GENOVEVA   
344  27-30744939-6  2019   12              FERRERO           GENOVEVA   

                                        reparticion  asignacion_por_cargo_i  \
21   SECR Administración de Seguridad y Emergencias               204017.27   
53   SECR Administración de Seguridad y Emergencias               204017.27   
85   SECR Administración de Seguridad y Emergencias               210061.14   
116  SECR Administración de Seguridad y Emergencias               226866.41   
147  SECR Administración de Seguridad y Emergencias               226866.41   
178  SECR Administración de Seguridad y Emergencias               226866.41   
209  SECR Administración de Seguridad y Emergencias               239470.36   
240  SECR Administración de Seguridad y Emergencias               239470.36   
271  SECR Administración de Seguridad y Emergencias               249972.87   
302  SECR Administración de Seguridad y Emergencias               249972.87   
333  SECR Administración de Seguridad y Emergencias               249972.87   
344  SECR Administración de Seguridad y Emergencias                74991.86   

     aguinaldo_ii  total_salario_bruto_i_+_ii observaciones  
21           0.00                   204017.27           NaN  
53           0.00                   204017.27           NaN  
85           0.00                   210061.14           NaN  
116          0.00                   226866.41           NaN  
147          0.00                   226866.41           NaN  
178     113433.21                   340299.62           NaN  
209          0.00                   239470.36           NaN  
240          0.00                   239470.36           NaN  
271          0.00                   249972.87           NaN  
302          0.00                   249972.87           NaN  
333          0.00                   249972.87           NaN  
344     110404.68                   185396.54  baja al 9/12

## Otros análisis descriptivos

Pandas viene con algunas funciones built-in para ayudar al análisis descriptivo.

## Para las variables numéricas

In [ ]:
df.describe()

anio         mes  asignacion_por_cargo_i   aguinaldo_ii  \
count   385.0  385.000000              385.000000     385.000000   
mean   2019.0    6.631169           234234.368000   20282.084883   
std       0.0    3.539077            35043.160085   45248.840725   
min    2019.0    1.000000            74991.860000       0.000000   
25%    2019.0    4.000000           224516.620000       0.000000   
50%    2019.0    7.000000           239470.360000       0.000000   
75%    2019.0   10.000000           249972.870000       0.000000   
max    2019.0   12.000000           275089.750000  170855.560000   

       total_salario_bruto_i_+_ii  
count                  385.000000  
mean                254516.452883  
std                  51434.987870  
min                 185396.540000  
25%                 224516.620000  
50%                 245811.620000  
75%                 263531.980000  
max                 445945.310000

## Para las variables categóricas


In [ ]:
df['reparticion'].value_counts()

reparticion
SECR Desarrollo Ciudadano                                         13
Ministerio de Salud                                               13
SECR Justicia y Seguridad                                         13
SECR de Medios                                                    13
Jefe de Gobierno                                                  12
SECR Legal y Técnica                                              12
SS Contenidos                                                     12
Ente de Turismo Ley Nº 2627                                       12
Consejo de los Derechos de Niñas, Niños y Adoles - Presidencia    12
Vicejefatura de Gobierno                                          12
UPE Transferencia de Funciones y Facultades en Materia de Seg.    12
SECR Administración de Seguridad y Emergencias                    12
SECR Integración Social y Urbana                                  12
SECR Integración Social Para Personas Mayores                     12
SECR Planificación, Ev

### Ejercicio: Volviendo al ejemplo del Titanic.

1) ¿Cuál era la edad promedio de los pasajeros de cada clase (Pclass)?

2) ¿Cuál fue la tarifa que pagaron en promedio los hombres? ¿Y las mujeres?

3) ¿Cuánto pagaron en total los pasajeros de primera clase para subir al Titanic? ¿Y los de tercera?

4) ¿Cuántos pasajeros había en cada tipo de clase?

In [ ]:
df_titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
# 1) ¿Cuál era la edad promedio de los pasajeros de cada clase (Pclass)?
df_titanic.groupby('Pclass') ['Age'].mean()


Pclass
1    38.233441
2    29.877630
3    25.140620
Name: Age, dtype: float64

In [ ]:
# 2) ¿Cuál fue la tarifa que pagaron en promedio los hombres? ¿Y las mujeres?
df_titanic.groupby('Sex') ['Fare'].mean()
#agrupo por los diferentes sexos (female or male). Despues saco el promedio de la tarifa (Fare)

Sex
female    44.479818
male      25.523893
Name: Fare, dtype: float64

In [ ]:
# 3) ¿Cuánto pagaron en total los pasajeros de primera clase para subir al Titanic? ¿Y los de tercera?
df_titanic.groupby('Pclass') ['Fare'].sum()

Pclass
1    18177.4125
2     3801.8417
3     6714.6951
Name: Fare, dtype: float64

*Los pasajeros de primera clase pagaron un total de 18177, 41 usd mientras que los de tercera pagaron 6714,69  usd*

In [ ]:
# 4) ¿Cuántos pasajeros había en cada tipo de clase?
df_titanic['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [ ]:
df_titanic['Pclass'].value_counts().sum()

891

## Ordenar por columnas y limitar la cantidad de resultados

Otra forma de resolver el problema de encontrar el mayor y el menos es con el método sort_values. Este método puede recibir un valor único (nombre de columna) o una lista (con varias columnas) y un orden asc o desc. Por default el orden es asc.

Si combinamos el ordenamiento con el método head() para limitar la cantidad de resultados, podemos encontrar los N primeros.

In [ ]:
# Recordemos cómo abrir la documentación de un método
df.sort_values?

In [ ]:
# Calculamos el máximo
df.sort_values('total_salario_bruto_i_+_ii',ascending=False).head(1)

cuil  anio  mes funcionario_apellido funcionario_nombre  \
374  20-13872301-2  2019   12             ASTARLOA      GABRIEL MARIA   

                                          reparticion  asignacion_por_cargo_i  \
374  Procuración General de la Ciudad de Buenos Aires               275089.75   

     aguinaldo_ii  total_salario_bruto_i_+_ii observaciones  
374     170855.56                   445945.31           NaN

In [ ]:
# Calculamos el mínimo
df.sort_values('total_salario_bruto_i_+_ii').head(1)

cuil  anio  mes funcionario_apellido funcionario_nombre  \
344  27-30744939-6  2019   12              FERRERO           GENOVEVA   

                                        reparticion  asignacion_por_cargo_i  \
344  SECR Administración de Seguridad y Emergencias                74991.86   

     aguinaldo_ii  total_salario_bruto_i_+_ii observaciones  
344     110404.68                   185396.54  baja al 9/12

## Anexo: volviendo al tema de la vectorización

¿Por qué es tan importante trabajar con Pandas y no con funciones escritas por nosotros en Python nativo y que procesen los datos dentro de un for loop?

Por un lado está la comodidad. Hay mucha funcionalidad que ya está desarrollada en Pandas. Existen funciones que resuelven muchos de los problemas clásicos de manipular datos: agrupar, sumarizar, sacar estadísticas, filtrar, etc. Pero además hay una razón de performance.

Veamos una demostración de que vectorizar es más eficiente. Vamos a crear dos listas de 1.000.000 de números aleatorios cada una y vamos a tratar de multiplicar elemento por elemento con pandas y sin pandas:



In [ ]:
# creamos dos listas sin usar pandas
lista1 = list(np.random.randint(1, 100, 1000000)) #creamos una lista de 1 a 1 millon saltando de a 100
lista2 = list(np.random.randint(1, 100, 1000000)) #creamos una lista de 1 a 1 millon saltando de a 100

In [ ]:
%%timeit
for x,y in zip(lista1,lista2):
    x * y

# un loop que para cada numero de la lista 1 se multilica por el de la lista 2

132 ms ± 85.8 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


Ahora probemos hacer lo mismo con dos series de Pandas

In [ ]:
serie1 = pd.Series(lista1)
serie2 = pd.Series(lista2)

In [ ]:
%%timeit
resultado = serie1 * serie2 #es lo mismo que el loop for

2.09 ms ± 276 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


Conclusión: la operación vectorizada es <strong> más de 70 veces más rápida.</strong>

In [ ]:
# para calcular utilizacion de recursos podemos usar %%time o %%timeit
%%time
resultado = serie1 * serie2

CPU times: user 2.79 ms, sys: 0 ns, total: 2.79 ms
Wall time: 2.83 ms
